typing/subst.ml

```diff
@@ -134,11 +134,38 @@ let newpersty desc =
   { desc; level = generic_level; scope = Btype.lowest_level; id = !new_id }
 
 (* ensure that all occurrences of 'Tvar None' are physically shared *)
-let tvar_none = Tvar None
-let tunivar_none = Tunivar None
+let tvar_none_any = Tvar { name = None; layout = Type_layout.any }
+let tvar_none_imm = Tvar { name = None; layout = Type_layout.immediate }
+let tvar_none_imm64 = Tvar { name = None; layout = Type_layout.immediate64 }
+let tvar_none_val = Tvar { name = None; layout = Type_layout.value }
+let tvar_none_void = Tvar { name = None; layout = Type_layout.void }
+
+let tunivar_none_any = Tunivar { name = None; layout = Type_layout.any }
+let tunivar_none_imm = Tunivar { name = None; layout = Type_layout.immediate }
+let tunivar_none_imm64 =
+  Tunivar { name = None; layout = Type_layout.immediate64 }
+let tunivar_none_val = Tunivar { name = None; layout = Type_layout.value }
+let tunivar_none_void = Tunivar { name = None; layout = Type_layout.void}
+
 let norm = function
-  | Tvar None -> tvar_none
-  | Tunivar None -> tunivar_none
+  | (Tvar { name = None; layout }) as t -> begin
+      match Type_layout.repr layout with
```

Comment:
I guess your concern is the cost of `repr` vs the savings of sharing the layouts?

At the moment it's a bit tricky to remove repr because of the rework that made it so we can't forget it :).  But for what it's worth I did a rough measurement of file sizes within the compiler when I put this optimization in and found that the difference was negligible, so I don't think this is buying us much.